tool/teleport/common/sftp.go

```diff
@@ -0,0 +1,326 @@
+		if !ok {
+			// We don't care about this type of SFTP request, move on
+			return
+		}
+
+		sftpEvents <- event
+	}))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Start a goroutine to marshal and send audit events to the parent
+	// process to avoid blocking the SFTP connection on event handling
+	done := make(chan struct{})
+	go func() {
+		for event := range sftpEvents {
+			eventBytes, err := json.Marshal(event)
+			if err != nil {
+				log.WithError(err).Warn("Failed to marshal SFTP event.")
+			} else {
+				// Append a NULL byte so the parent process will know where
+				// this event ends
+				eventBytes = append(eventBytes, 0x0)
+				_, err = io.Copy(auditFile, bytes.NewReader(eventBytes))
+				if err != nil {
+					log.WithError(err).Warn("Failed to send SFTP event to parent.")
+				}
+			}
+		}
+
+		close(done)
+	}()
+
+	serveErr := sftpSrv.Serve()
+	if errors.Is(serveErr, io.EOF) {
+		serveErr = nil
+	} else {
+		serveErr = trace.Wrap(serveErr)
+	}
+
+	// Wait until event marshalling goroutine is finished
+	close(sftpEvents)
+	<-done
+
+	return trace.NewAggregate(serveErr, sftpSrv.Close())
+}
+
+func handleSFTPEvent(reqPacket sftp.RequestPacket, path string, opErr error) (*apievents.SFTP, bool) {
+	event := &apievents.SFTP{
+		Metadata: apievents.Metadata{
+			Type: events.SFTPEvent,
+			Time: time.Now(),
+		},
+	}
+
+	switch p := reqPacket.(type) {
+	case *sftp.OpenPacket:
+		if opErr == nil {
+			event.Code = events.SFTPOpenCode
+		} else {
+			event.Code = events.SFTPOpenFailureCode
+		}
+		event.Action = apievents.SFTPAction_OPEN
+		event.Path = p.Path
+		event.Flags = p.Pflags
+	case *sftp.ClosePacket:
+		if opErr == nil {
+			event.Code = events.SFTPCloseCode
+		} else {
+			event.Code = events.SFTPCloseFailureCode
+		}
+		event.Action = apievents.SFTPAction_CLOSE
+		event.Path = path
+	case *sftp.ReadPacket:
+		if opErr == nil {
+			event.Code = events.SFTPReadCode
+		} else {
+			event.Code = events.SFTPReadFailureCode
+		}
+		event.Action = apievents.SFTPAction_READ
+		event.Path = path
+	case *sftp.WritePacket:
+		if opErr == nil {
+			event.Code = events.SFTPWriteCode
+		} else {
+			event.Code = events.SFTPWriteFailureCode
+		}
+		event.Action = apievents.SFTPAction_WRITE
+		event.Path = path
+	case *sftp.LstatPacket:
+		if opErr == nil {
+			event.Code = events.SFTPLstatCode
+		} else {
+			event.Code = events.SFTPLstatFailureCode
+		}
+		event.Action = apievents.SFTPAction_LSTAT
+		event.Path = p.Path
+	case *sftp.FstatPacket:
+		if opErr == nil {
+			event.Code = events.SFTPFstatCode
+		} else {
+			event.Code = events.SFTPFstatFailureCode
+		}
+		event.Action = apievents.SFTPAction_FSTAT
+		event.Path = path
+	case *sftp.SetstatPacket:
+		if opErr == nil {
+			event.Code = events.SFTPSetstatCode
+		} else {
+			event.Code = events.SFTPSetstatFailureCode
+		}
+		event.Action = apievents.SFTPAction_SETSTAT
+		event.Path = p.Path
+		event.Attributes = unmarshalSFTPAttrs(p.Flags, p.Attrs.([]byte))
```

Comment:
I see that `Attributes` are not always set. The same with `Flags` too. What is the rule? Can we make this code a little bit more generic?

Reply:
I wanted to avoid creating many audit events for every type of SFTP request, so I just created one containing all the fields necessary to describe every SFTP request. The way I'm handling it now is if the SFTP request has a field set (ie `Open` uses `Flags`) then I will set that in the SFTP audit event.

I'm honestly not sure how I can make this code more generic as I have to set a specific code and action for each request, and I have to use the path provided from the callback instead of the request if the request uses a file handle.